app/controllers/api/connect/v4/systems/products_controller.rb

```diff
@@ -15,4 +15,26 @@ def destroy
     raise ActionController::TranslatedError.new(N_('%s is not yet activated on the system.'), @product.name)
   end
 
+  def synchronize
+    products = params.require(:products).map do |product_params|
+      @system.products.find_by(
+        identifier: product_params[:identifier],
+        version: product_params[:version],
+        arch: product_params[:arch]
+      )
+    end
+
+    (@system.products - products).each do |product|
+      ActiveRecord::Base.transaction do
```

Comment:
Why would you wrap activations in separate transactions each, one by one?
Is it because of some `before_destroy` callbacks?

Otherwise maybe let's just do an atomic `destroy_all` query instead each time?
```ruby
@system.activations.includes(:service).where('services.product_id' => product.id).destroy_all
```

Reply:
Hmm, yeah, doesn't make much sense as it is.
I guess it would make more sense to wrap the whole `.each` loop into one big transaction, so that you don't end up with half of the activations destroyed in case of an error.
(Also copypasta)